types/underscore/index.d.ts

```diff
@@ -890,15 +929,14 @@ declare module _ {
         compact<T>(array: _.List<T | null | undefined | false | "" | 0> | null | undefined): T[]
 
         /**
-        * Flattens a nested array (the nesting can be to any depth). If you pass shallow, the array will
-        * only be flattened a single level.
-        * @param array The array to flatten.
-        * @param shallow If true then only flatten one level, optional, default = false.
-        * @return `array` flattened.
-        **/
-        flatten(
-            array: _.List<any>,
-            shallow?: boolean): any[];
+         * Flattens a nested array (the nesting can be to any depth). If you pass shallow, the array will
+         * only be flattened a single level.
+         * @param list The array to flatten.
+         * @param shallow If true then only flatten one level, optional, default = false.
+         * @returns The flattened list.
+         **/
+        flatten<T>(list: List<T>, shallow?: false): DeepFlattenedList<T>[];
```

Comment:
Would you mind providing an example of such a situation?

Reply:
My tests check one-, two-, three-, and four-dimension arrays and even a type union of arrays of different dimensions, so I'm not sure what scenarios I might be missing here. Arrays of four dimensions or more will still result in `any[]`.

Reply:
My mistake. There were at least several minutes, possible even half an hour, between me reading the definition of `DeepFlattenedList` and this `flatten` type. At this point, I remembered that `DeepFlattenedList` was limited to two levels of nesting, but I did not remember that there was a fallback to `any`. Re-reading that definition now, I see that there is.

So I'm not worried anymore that this will give incorrect results. On the other hand, I'm not entirely sure how useful this is. This is only going to improve type inference compared to the old situation if nesting is shallower than three levels, at the expense of TS working a sweat even when nesting is three levels or deeper.

Reply:
I don't think it'll be a huge sweat, though, and it will at least be a constant sweat as well after the threshold is reached.

My biggest concern here is that I think it's common for folks to call `flatten` without any parameters even when what they're technically getting is a shallow flatten of a two-dimensional array, so I'd be fine with reducing the amount of recursion I'm doing if you'd prefer to only handle that common case, but I really really don't want to go all the way back to `any[]`.

Reply:
Also `_ChainOfArrays` was put in place explicitly to handle the common case of flattening a two-dimensional array produced by `map` using a "deep" flatten, so I should ideally keep support of at least two dimensions in place to continue covering that case as well.

Reply:
Alright!